File: frontend/src/api/pricing.ts
import apiFetch from "./client";

export interface LinePricingOut {
  product_id: string;
  quantity: number;
  base_price: string;
  company_pct: string;
  price_after_company: string;
  volume_pct: string;
  final_unit_price: string;
  effective_pct: string;
  line_total: string;
}

export interface QuoteOut {
  lines: LinePricingOut[];
  subtotal_base: string;
  total_discount: string;
  grand_total: string;
}

export interface PricingTierOut {
  min_quantity: number;
  discount_percentage: string;
  unit_price: string;
}

export interface ProductPricingOut {
  base_price: string;
  company_discount_percentage: string;
  unit_price: string;
  tiers: PricingTierOut[];
}

export interface QuoteItem {
  product_id: string;
  quantity: number;
}

export const getProductPricing = (slug: string, mode: string) =>
  apiFetch<ProductPricingOut>(`/pricing/product/${slug}?mode=${mode}`, {
    method: "GET",
  });

export const quoteCart = (items: QuoteItem[], mode: string) =>
  apiFetch<QuoteOut>("/pricing/quote", {
    method: "POST",
    body: { items, mode },
  });

File: backend/app/schemas/pricing.py
from decimal import Decimal
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field


class QuoteItemIn(BaseModel):
    product_id: UUID
    quantity: int = Field(ge=1)


class QuoteRequest(BaseModel):
    items: list[QuoteItemIn]
    mode: Literal["COMPANY", "PRIVATE"] = "PRIVATE"


class LinePricingOut(BaseModel):
    product_id: UUID
    quantity: int
    base_price: Decimal
    company_pct: Decimal
    price_after_company: Decimal
    volume_pct: Decimal
    final_unit_price: Decimal
    effective_pct: Decimal
    line_total: Decimal


class QuoteOut(BaseModel):
    lines: list[LinePricingOut]
    subtotal_base: Decimal
    total_discount: Decimal
    grand_total: Decimal


class PricingTierOut(BaseModel):
    min_quantity: int
    discount_percentage: Decimal
    unit_price: Decimal


class ProductPricingOut(BaseModel):
    base_price: Decimal
    company_discount_percentage: Decimal
    unit_price: Decimal
    tiers: list[PricingTierOut]
